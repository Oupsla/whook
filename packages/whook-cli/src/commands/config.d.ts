import { LogService } from 'common-services';
import { CONFIGSService } from '@whook/whook';
import {
  PromptArgs,
  WhookCommandDefinition,
  WhookCommandHandler,
} from '../services/promptArgs';
export declare const definition: WhookCommandDefinition;
declare const _default: typeof initConfigCommand;
export default _default;
declare function initConfigCommand({
  CONFIGS,
  promptArgs,
  log,
}: {
  CONFIGS: CONFIGSService;
  promptArgs: PromptArgs;
  log?: LogService;
}): Promise<WhookCommandHandler>;
