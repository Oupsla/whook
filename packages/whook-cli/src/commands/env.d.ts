import { WhookCommandDefinition, PromptArgs } from '../services/promptArgs';
import { ENVService } from '@whook/whook';
import { LogService } from 'common-services';
export declare const definition: WhookCommandDefinition;
declare const _default: typeof initEnvCommand;
export default _default;
declare function initEnvCommand({
  ENV,
  promptArgs,
  log,
}: {
  ENV: ENVService;
  promptArgs: PromptArgs;
  log?: LogService;
}): Promise<() => Promise<void>>;
