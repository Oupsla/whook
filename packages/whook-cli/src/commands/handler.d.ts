import { Injector } from 'knifecycle';
import {
  WhookCommandDefinition,
  PromptArgs,
  WhookCommandHandler,
} from '../services/promptArgs';
import { WhookHandler } from '@whook/whook';
import { LogService } from 'common-services';
export declare const definition: WhookCommandDefinition;
declare const _default: typeof initHandlerCommand;
export default _default;
declare function initHandlerCommand({
  $injector,
  log,
  promptArgs,
}: {
  $injector: Injector<WhookHandler>;
  log?: LogService;
  promptArgs: PromptArgs;
}): Promise<WhookCommandHandler>;
