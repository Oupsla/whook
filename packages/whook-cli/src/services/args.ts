import { autoService, options } from 'knifecycle';
import parseArgs from 'yargs-parser';
import { LogService } from 'common-services';
import { WhookArgsTypes } from '..';

export default options({ singleton: true }, autoService(initArgs));

interface WhookCommandNamedArgs {
  [name: string]: WhookArgsTypes;
}
interface WhookCommandArgsRest {
  _?: string[];
}
export type WhookCommandArgs = WhookCommandNamedArgs & WhookCommandArgsRest;

async function initArgs({
  ARGS,
  log,
}: {
  ARGS: string[];
  log: LogService;
}): Promise<WhookCommandArgs> {
  const args = parseArgs(ARGS.slice(2));

  log('debug', 'ðŸ›  - Parsed args:', args);
  return args;
}
