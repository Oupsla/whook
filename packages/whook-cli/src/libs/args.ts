import Ajv from 'ajv';
import YError from 'yerror';
import { WhookCommandDefinitionArguments } from '../services/promptArgs';
import { WhookCommandArgs } from '../services/args';

// TODO: Add ajv human readable error builder
export function readArgs(
  schema: WhookCommandDefinitionArguments,
  args: WhookCommandArgs,
): WhookCommandArgs {
  const ajv = new Ajv({
    coerceTypes: true,
    useDefaults: true,
    strictDefaults: true,
    strictKeywords: true,
  });
  const validator = ajv.compile(schema);

  const {
    _: [, ...listedArgs],
    ...namedArgs
  } = args;
  const cleanedArgs = {
    ...namedArgs,
    ...(listedArgs.length ? { _: listedArgs } : {}),
  } as WhookCommandArgs;

  validator(cleanedArgs);

  if ((validator.errors || []).length) {
    throw new YError('E_BAD_ARGS', validator.errors);
  }

  return cleanedArgs;
}
