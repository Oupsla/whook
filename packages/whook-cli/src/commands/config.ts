import { extra, autoService } from 'knifecycle';
import { readArgs } from '../libs/args';
import YError from 'yerror';
import miniquery from 'miniquery';
import { LogService } from 'common-services';
import { noop, CONFIGSService } from '@whook/whook';
import {
  PromptArgs,
  WhookCommandDefinition,
  WhookCommandHandler,
} from '../services/promptArgs';

export const definition: WhookCommandDefinition = {
  description: 'A simple program that returns the queryed config value',
  example: `whook config --name MYSQL --query 'auth.username' --default root`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name'],
    properties: {
      name: {
        description: 'Configuration name',
        type: 'string',
      },
      query: {
        description: 'Property to pickup in config (uses `miniquery`)',
        type: 'string',
      },
      default: {
        description: 'Provide a default value',
        type: 'string',
      },
    },
  },
};

export default extra(definition, autoService(initConfigCommand));

async function initConfigCommand({
  CONFIGS,
  promptArgs,
  log = noop,
}: {
  CONFIGS: CONFIGSService;
  promptArgs: PromptArgs;
  log?: LogService;
}): Promise<WhookCommandHandler> {
  return async () => {
    const { name, query, default: defaultValue } = readArgs(
      definition.arguments,
      await promptArgs(),
    ) as { name: string; query: string; default: unknown };

    if ('undefined' === typeof CONFIGS[name]) {
      log('error', `No config found for ${name}`);
      if ('undefined' === typeof defaultValue) {
        throw new YError('E_NO_CONFIG', name);
      }
    }

    const results = query ? miniquery(query, [CONFIGS[name]]) : [CONFIGS[name]];

    if (!results.length) {
      log('error', `Could not find any results for ${query}`);
      if ('undefined' === typeof defaultValue) {
        throw new YError('E_NO_RESULT', name, query);
      }
    }

    if (results.length > 1) {
      log(
        'error',
        `Got ${results.length} results for the "${query}" query, picking-up the first one.`,
      );
    }

    log(
      'info',
      `${JSON.stringify(results.length ? results[0] : defaultValue)}`,
    );
  };
}
