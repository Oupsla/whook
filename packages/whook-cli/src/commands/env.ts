import { extra, autoService } from 'knifecycle';
import { readArgs } from '../libs/args';
import YError from 'yerror';
import { WhookCommandDefinition, PromptArgs } from '../services/promptArgs';
import { ENVService, noop } from '@whook/whook';
import { LogService } from 'common-services';

export const definition: WhookCommandDefinition = {
  description: 'A command printing env values',
  example: `whook env --name NODE_ENV --default "default value"`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name'],
    properties: {
      name: {
        description: 'Environment variable name to pick-up',
        type: 'string',
      },
      default: {
        description: 'Provide a default value',
        type: 'string',
      },
    },
  },
};

export default extra(definition, autoService(initEnvCommand));

async function initEnvCommand({
  ENV,
  promptArgs,
  log = noop,
}: {
  ENV: ENVService;
  promptArgs: PromptArgs;
  log?: LogService;
}) {
  return async () => {
    const { name, default: defaultValue } = readArgs(
      definition.arguments,
      await promptArgs(),
    ) as { name: string; default: unknown };

    if (
      'undefined' === typeof ENV[name] &&
      'undefined' === typeof defaultValue
    ) {
      throw new YError('E_NO_ENV_VALUE', name);
    }

    log('info', `${ENV[name] || defaultValue}`);
  };
}
