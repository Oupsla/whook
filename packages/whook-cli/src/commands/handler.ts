import { extra, autoService, Injector, Parameters } from 'knifecycle';
import YError from 'yerror';
import { readArgs } from '../libs/args';
import {
  WhookCommandDefinition,
  PromptArgs,
  WhookCommandHandler,
} from '../services/promptArgs';
import { noop, WhookHandler, WhookResponse } from '@whook/whook';
import { LogService } from 'common-services';

export const definition: WhookCommandDefinition = {
  description: 'Runs the given server handler for testing purpose',
  example: `whook handler --name getPing --parameters '{}'`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name', 'parameters'],
    properties: {
      name: {
        description: 'Handler to invoke',
        type: 'string',
      },
      parameters: {
        description: 'Parameters to invoke the handler with',
        type: 'string',
        default: '{}',
      },
    },
  },
};

export default extra(definition, autoService(initHandlerCommand));

async function initHandlerCommand({
  $injector,
  log = noop,
  promptArgs,
}: {
  $injector: Injector<WhookHandler>;
  log?: LogService;
  promptArgs: PromptArgs;
}): Promise<WhookCommandHandler> {
  return async () => {
    const { name: handlerName, parameters: handlerParameters } = readArgs(
      definition.arguments,
      await promptArgs(),
    ) as { name: string; parameters: string };
    let parsedParameters: Parameters;

    try {
      parsedParameters = JSON.parse(handlerParameters || '{}');
    } catch (err) {
      throw YError.wrap(err, 'E_BAD_PARAMETERS', handlerParameters);
    }

    log('debug', 'handler', handlerName);
    log('debug', 'parameters', parsedParameters);

    // Maybe infer and check command definition from handler definition
    // with ajv or else

    try {
      const handler: WhookHandler<any, any, any> = (
        await $injector([handlerName])
      )[handlerName];
      const response: WhookResponse = await handler(
        parsedParameters,
        undefined,
      );

      log('info', JSON.stringify(response, null, 2));
    } catch (err) {
      log('error', 'Got an error while running the handler.');
      log('stack', err.stack);
    }
  };
}
