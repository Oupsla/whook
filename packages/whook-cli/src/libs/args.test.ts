import { readArgs } from './args';
import { definition as handlerCommandDefinition } from '../commands/handler';
import YError from 'yerror';

describe('readArgs', () => {
  it('should work with no args', () => {
    const args = {
      _: ['whook'],
    };

    readArgs(
      {
        type: 'object',
      } as any,
      args as any,
    );

    expect({
      args,
    }).toMatchSnapshot();
  });

  it('should work with named args', () => {
    const args = {
      _: ['whook'],
      name: 'getPing',
      parameters: '{}',
    };

    readArgs(handlerCommandDefinition.arguments, args as any);

    expect({
      args,
    }).toMatchSnapshot();
  });

  it('should work with listed args', () => {
    const args = {
      _: ['whook', 'hey'],
    };

    readArgs(
      {
        type: 'object',
        additionalProperties: false,
        properties: {
          _: {
            type: 'array',
            maxItems: Infinity,
          },
        },
      } as any,
      args as any,
    );

    expect({
      args,
    }).toMatchSnapshot();
  });

  it('should report named args errors', () => {
    const args = {
      _: ['whook'],
      parameters: '{}',
    };

    try {
      readArgs(handlerCommandDefinition.arguments, args as any);
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        args,
        errorCode: err.code,
        errorParams: err.params,
      }).toMatchSnapshot();
    }
  });
});
