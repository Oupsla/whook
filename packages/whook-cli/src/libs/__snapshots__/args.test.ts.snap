// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readArgs should report named args errors 1`] = `
Object {
  "args": Object {
    "_": Array [
      "whook",
    ],
    "parameters": "{}",
  },
  "errorCode": "E_BAD_ARGS",
  "errorParams": Array [
    Array [
      Object {
        "dataPath": "",
        "keyword": "required",
        "message": "should have required property 'name'",
        "params": Object {
          "missingProperty": "name",
        },
        "schemaPath": "#/required",
      },
    ],
  ],
}
`;

exports[`readArgs should work with listed args 1`] = `
Object {
  "args": Object {
    "_": Array [
      "whook",
      "hey",
    ],
  },
}
`;

exports[`readArgs should work with named args 1`] = `
Object {
  "args": Object {
    "_": Array [
      "whook",
    ],
    "name": "getPing",
    "parameters": "{}",
  },
}
`;

exports[`readArgs should work with no args 1`] = `
Object {
  "args": Object {
    "_": Array [
      "whook",
    ],
  },
}
`;
